src/commands/preview/proxy.rs

```diff
@@ -20,104 +18,175 @@ use url::Url;
+    is_https: bool,
+}
+
+impl ProxyConfig {
+    pub fn new(
+        host: Option<&str>,
+        ip: Option<&str>,
+        port: Option<&str>,
+    ) -> Result<Self, failure::Error> {
+        let port: &str = match port {
+            Some(port) => port,
+            None => "8000",
+        };
+
+        let try_address = match ip {
+            Some(ip) => format!("{}:{}", ip, port),
+            None => format!("localhost:{}", port),
+        };
+
+        let mut address_iter = try_address.to_socket_addrs()?;
+
+        let listening_address = match address_iter.next() {
+            Some(ip) => Ok(ip),
+            None => Err(format_err!("Could not parse address {}", try_address)),
+        }?;
+
+        let host: String = match host {
+            Some(host) => host.to_string(),
+            None => "https://example.com".to_string(),
+        };
+
+        let parsed_url = match Url::parse(&host) {
+            Ok(host) => Ok(host),
+            Err(_) => Url::parse(&format!("https://{}", host)),
+        }?;
+
+        let scheme: &str = parsed_url.scheme();
+        if scheme != "http" && scheme != "https" {
+            failure::bail!("Your host scheme must be either http or https")
+        }
+        let is_https = scheme == "https";
+
+        let host = match parsed_url.host_str() {
+            Some(host_str) => Ok(host_str.to_string()),
+            None => Err(format_err!(
+                "Invalid host, accepted formats are http://example.com or example.com"
+            )),
+        }?;
+
+        let proxy = ProxyConfig {
+            listening_address,
+            host,
+            is_https,
+        };
+
+        Ok(proxy)
+    }
+
+    fn get_listening_address_as_str(&self) -> String {
+        self.listening_address
+            .to_string()
+            .replace("[::1]", "localhost")
+    }
+}
+
+pub async fn proxy(
+    target: Target,
     user: Option<GlobalUser>,
     host: Option<&str>,
     port: Option<&str>,
+    ip: Option<&str>,
 ) -> Result<(), failure::Error> {
-    commands::build(&target)?;
-    let port: u16 = match port {
-        Some(port) => port.to_string().parse(),
-        None => Ok(3000),
-    }?;
-    let host: String = match host {
-        Some(host) => host.to_string(),
-        None => "https://example.com".to_string(),
-    };
-    let parsed_url = match Url::parse(&host) {
-        Ok(host) => Ok(host),
-        Err(_) => Url::parse(&format!("https://{}", host)),
-    }?;
-    let scheme: &str = parsed_url.scheme();
-    if scheme != "http" && scheme != "https" {
-        failure::bail!("Your host must be either http or https")
+    let proxy_config = ProxyConfig::new(host, ip, port)?;
+    let https = HttpsConnector::new().expect("TLS initialization failed");
+    let client = Client::builder().build::<_, Body>(https);
+
+    let preview_id = get_preview_id(target, user, &proxy_config)?;
+    let listening_address = &proxy_config.listening_address.clone();
```

Comment:
doing `&foo.clone()` always feels a bit weird; I would drop the `&` here and use the `&` at the call site below personal preference.